### PR TITLE
Fix flow comments plugin issues

### LINKED ISSUE #10321: Flow comments plugin removes generic type arguments of the right hand side of the `extends` keyword
## Bug Report

**Current Behavior**
`@babel/plugin-transform-flow-comments` removes generic type arguments of the right hand side of the `extends` keyword

```js
class A extends B {}
```

**Input Code**
[REPL link](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=true&code_lz=MYGwhgzhAECC0FMAeAXBA7AJjAQgHgA0A-aAbwF8AoIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.5.5&externalPlugins=babel-plugin-syntax-jsx%406.18.0%2C%40babel%2Fplugin-transform-flow-comments%407.5.5)

```js
class A extends B<X> {}
```

**Expected behavior/code**

```js
class A extends B /*:: <X>*/ {}
```

**Environment**
- Babel version(s): v7.5.5
- How you are using Babel: Online REPL

**Possible Solution**
```js

  return {
    //...
    visitor: {
      //...
      ClassDeclaration(path) {
        if (path.parentPath.isFlow()) {
          return;
        }

        const { node } = path;
        if (node.superTypeParameters) {
          const superTypeParameters = path.get("superTypeParameters");
          const superClass = path.get("superClass");
          superClass.addComment("trailing", generateComment(superTypeParameters, node));
          superTypeParameters.remove();
        }
      },
      //...
    },
  };
```

### LINKED ISSUE #10323: Flow comments plugin moves the generic type arguments of the left hand side of the `extends` keyword to its right side
## Bug Report

**Current Behavior**
`@babel/plugin-transform-flow-comments` moves the generic type arguments of the left hand side of the `extends` keyword to its right side

```js
class A extends B /*:: <X>*/ {}
```

**Input Code**
[REPL link](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=true&code_lz=MYGwhgzhAECCA8ANAfNApgDwC5oHYBMYAhaAbwF8AoIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.5.5&externalPlugins=%40babel%2Fplugin-transform-flow-comments%407.5.5)

```js
class A<X> extends B {}
```

**Expected behavior/code**

```js
class A /*:: <X>*/ extends B {}
```

**Environment**
- Babel version(s): v7.5.5
- How you are using Babel: Online REPL

**Possible Solution**
```js

  return {
    //...
    visitor: {
      //...
      ClassDeclaration(path) {
        if (path.parentPath.isFlow()) {
          return;
        }

        const { node } = path;
        if (node.typeParameters) {
          const typeParameters = path.get("typeParameters");
          const id = path.get("id");
          id.addComment("trailing", generateComment(typeParameters, node));
          typeParameters.remove();
        }
      },
      //...
    },
  };
```

### LINKED ISSUE #10324: Flow comments change order of comments at beginning of file
## Bug Report

**Current Behavior**
`@babel/plugin-transform-flow-comments` change order of comments at beginning of file when the file starts with a comment followed by a flow statement

```js
/*:: import type {A} from 'a';*/

/**/
X;
```

**Input Code**
[REPL link](https://babeljs.io/repl#?babili=false&browsers=&build=&builtIns=false&spec=false&loose=true&code_lz=PQKhwKASwWwBwPYCcAuACFBPOBTNBvAQQF80AzJBGNAcgEMaBuCADSA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=&prettier=false&targets=&version=7.5.5&externalPlugins=%40babel%2Fplugin-transform-flow-comments%407.5.5)

```js
/**/
import type {A} from 'a';
X
```

**Expected behavior/code**

```js
/**/
/*:: import type {A} from 'a';*/
X;
```

**Environment**
- Babel version(s): v7.5.5
- How you are using Babel: Online REPL

**Possible Solution**
```js
  //...
  function attachComment(path, comment) {
    let attach = path.getPrevSibling();
    let where = "trailing";
    if (!attach.node) {
      attach = path.getNextSibling();
      where = "leading";
    }
    if (!attach.node) {
      attach = path.parentPath;
      where = "inner";
    }
    attach.addComment(where, comment);
    path.remove();
  }
  //...
```

### PR DESCRIPTION
**Pull request by zaygraveyard**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #10324, Fixes #10323, Fixes #10321
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  | No
| Minor: New Feature?      | No
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  | No
| License                  | MIT

Fix issues in `plugin-transform-flow-comments` where some comments might get lost or have their position changed.
And add support for correctly handling `extends` keyword in the `class` declaration.

### CODE REVIEW

**Comment by zaygraveyard** [comment]
Does anyone know why is the CI build failing?

**Comment by nicolo-ribaudo** [comment]
Yeah ignore CircleCI, it should be fixed by 0519650

**Comment by zaygraveyard** [comment]
@nicolo-ribaudo Thank you
Is it better to rebase this PR on top this fix?

**Comment by nicolo-ribaudo** [comment]
Yeah it would be appreciated but not needed

File: packages/babel-plugin-transform-flow-comments/src/index.js [hunk 1]
```diff
@@ -6,15 +6,46 @@ import generateCode from "@babel/generator";
...
+   * to it.
+   * @param  {Path} path - The path of the node to remove
+   */
+  function removeTypeAnnotation(path) {
+    const node = path.node;
+    const parentPath = path.parentPath;
+    const prevPath = path.getPrevSibling();
+    const nextPath = path.getNextSibling();
+    if (node.leadingComments && !prevPath.node) {
+      if (prevPath.node) {
```
[the remark this answers is not in the document]
Comment (zaygraveyard) [reply packages/babel-plugin-transform-flow-comments/src/index.js]: You are absolutely correct
I'm gonna add tests to check this part

[the remark this answers is not in the document]
Comment (zaygraveyard) [reply packages/babel-plugin-transform-flow-comments/src/index.js]: Looks like this change breaks other tests
I'll investigate more...

[the remark this answers is not in the document]
Comment (zaygraveyard) [reply packages/babel-plugin-transform-flow-comments/src/index.js]: Done in fb12b89

File: packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/class-type-parameter-extends-type-parameter/input.js [hunk 2]
```diff
@@ -0,0 +1 @@
+class Foo<T> extends Bar<R> {}
```
Comment (JLHwung) [inline packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/class-type-parameter-extends-type-parameter/input.js]: Could you add a test when there is a comment between class identifier and extends keyword?
```js
class Foo<T> /* inner */ extends Bar<R> {}
```

Comment (zaygraveyard) [reply packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/class-type-parameter-extends-type-parameter/input.js]: Thanks you for this case, unfortunately it was transformed into:
```js
class Foo
/*:: <T>*/
extends
/* inner */
Bar
/*:: <R>*/
{}
```
Which should have been
```js
class Foo
/*:: <T>*/
/* inner */
extends Bar
/*:: <R>*/
{}
```
Looking into it... :confused:

Comment (zaygraveyard) [reply packages/babel-plugin-transform-flow-comments/test/fixtures/flow-comments/class-type-parameter-extends-type-parameter/input.js]: Fixed in 422685f

**Pull request merged**